### PR TITLE
fix: append weighted destination only when weight is mentioned 

### PR DESCRIPTION
**Pull request by divyansh375**
As of now, in canary deployment with weighted experiment, the rollout throws nil pointer exception, when in an experiment step, one template contains weight and the other doesn't. For example: 

Screenshot 2023-04-19 at 5 03 25 PM

Hence handling a nil pointer check and also a UT verification for port name .

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/) with a list of types and scopes found [here](https://github.com/argoproj/argo-rollouts/blob/master/.github/workflows/pr-title-check.yml), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

**Comment by zachaller** [comment]
Can we also add a unit test for the nil pointer check?

**Comment by divyansh375** [comment]
Hi @zachaller thanks for reviewing, let me add that as well

**Comment by divyansh375** [comment]
Hi @zachaller apologies for taking so long to address your comment, I have made few changes in the existing UT itself (TestRolloutWithExperimentStep) to reduce code duplication and also, because the existing UT "TestRolloutWithExperimentStep" was not covering the desired code.. It was logging error but somehow the UT was passing.

File: rollout/trafficrouting_test.go [hunk 1]
```diff
@@ -292,18 +305,13 @@ func TestRolloutWithExperimentStep(t *testing.T) {
 			// make sure SetWeight was called with correct value
 			assert.Equal(t, int32(10), desiredWeight)
 			assert.Equal(t, int32(5), weightDestinations[0].Weight)
+			assert.Len(t, weightDestinations, 1)
 			assert.Equal(t, ex.Status.TemplateStatuses[0].ServiceName, weightDestinations[0].ServiceName)
 			assert.Equal(t, ex.Status.TemplateStatuses[0].PodTemplateHash, weightDestinations[0].PodTemplateHash)
 			return nil
 		})
 		f.fakeTrafficRouting.On("SetHeaderRoute", mock.Anything, mock.Anything).Return(nil)
-		f.fakeTrafficRouting.On("VerifyWeight", mock.Anything).Return(func(desiredWeight int32, weightDestinations ...v1alpha1.WeightDestination) error {
```
Comment (zachaller) [inline rollout/trafficrouting_test.go:300]: Why are these test cases being removed?

Comment (divyansh375) [reply rollout/trafficrouting_test.go:300]: @zachaller   because these assertions are already present here : https://github.com/argoproj/argo-rollouts/pull/2734/files#:~:text=with%20correct%20value-,assert.Equal(t%2C%20int32(10)%2C%20desiredWeight,%5B0%5D.PodTemplateHash%2C%20weightDestinations%5B0%5D.PodTemplateHash),-return%20nil

and also, these assertions were not being executed till now.. an exception was being thrown at the beginning of this UT but somehow this UT was not failing .. hence, I updated this existing UT itself so that these assertions are properly executed.

File: rollout/trafficrouting_test.go [hunk 2]
```diff
@@ -318,11 +326,7 @@ func TestRolloutWithExperimentStep(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (divyansh375) [reply rollout/trafficrouting_test.go:321]: because these assertions are already present here : https://github.com/argoproj/argo-rollouts/pull/2734/files#:~:text=with%20correct%20value-,assert.Equal(t%2C%20int32(10)%2C%20desiredWeight),-assert.Len

**Comment by divyansh375** [comment]
@zachaller please let me know if any further change is required before we can merge this

**Comment by divyansh375** [comment]
@leoluz please could you review?

**Comment by divyansh375** [comment]
@zachaller  please can you review

**Comment by divyansh375** [comment]
@zachaller please can you review

**Review by zachaller — APPROVED** [review]
